### PR TITLE
163518646  View and Edit Profile

### PR DESCRIPTION
**Pull request by niomwungeri-fabrice**
#### What does this PR do?
It lets a user view and edits their profile
#### Description of Task to be completed?
-  Create `Profile` components with React
-  Create `EditProfile` components with React
-  Create `Settings` components with React
- Integrate the components with redux for state management
- write unit tests for all elements used by those components 
#### How should this be manually tested?
- clone the repo and check out the following branch `ft-view-edit-profile-#23163518646`
- clone the repo and check out the following branch `ft-change-password-#23165713051` from our backend [repo](https://github.com/andela/Titan-Devs/pull/54) 
-  Refer to `.sample.env` to update your env file
- run `yarn start` 
-  Navigate `localhost:{port}/profiles/:{username}` for viewing a profile
 - Navigate `localhost:{port}/profiles/:{username}/edit` for editing a profile
 - Navigate `localhost:{port}/settings` for user settings
- run  `yarn test` to  run the unit tests
#### Any background context you want to provide?
The PR depends on the logged in user.
#### What are the relevant pivotal tracker stories?
#163518642 
#### Screenshots (if appropriate):
View page
Screen Shot 2019-04-14 at 13 40 07
Edit page
Screen Shot 2019-04-14 at 13 40 30

### CODE REVIEW

File: src/views/EditProfile.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        lastName,
+        bio,
+        phone,
+        address,
+        image
+      },
+      error,
+      message
+    } = this.props;
+    // const token = localStorage.getItem("token");
```
Comment (espoirMur) [inline src/views/EditProfile.js]: Can you please remove this comment here if it's not used?

File: yarn.lock [hunk 2]
```diff
@@ -671,7 +671,7 @@
     "@babel/plugin-transform-react-jsx-self" "^7.0.0"
```
Comment (espoirMur) [inline yarn.lock]: Bro! Please consider ignoring this file!

File: src/redux/actions/index.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      dispatch(setError(message));
+      dispatch(setLoading(false));
+    });
+};
+
+export const saveUpdatedUser = updatedProfile => dispatch =>
+  axios
+    .put(
+      `${process.env.API_URL}/profiles/${updatedProfile.username}`,
+      { profile: updatedProfile },
```
Comment (YvesIraguha) [inline src/redux/actions/index.js]: There are configurations for baseUrl and headers set in `login feature` you can use them to get the baseUrl and setting headers of the request.

**Review by espoirMur — COMMENTED** [review]
Nice Job Fab! Here are some keys points to consider : 

 - While reviewing this PR I was not able to edit my profile. I was getting a blank page with the edit button disabled... 
- Can you consider putting a default picture in case a user doesn't have any?
- It looks like I can edit my profile without being logged in, is it possible to restrict someone who is not logged to edit his profile?

File: src/__tests__/__actions__/loginActions.test.js [hunk 4]
```diff
@@ -14,7 +14,7 @@ import {
   LOGIN_INPUT_CHANGE
 } from "../../redux/actionTypes";
 
-const DEV_BASE_URL = "http://localhost:3000/api/v1";
+const DEV_BASE_URL = process.env.API_URL;
 let data;
```
Comment (YvesIraguha) [inline src/__tests__/__actions__/loginActions.test.js]: You can use the configurations set in `login feature` for getting the DEV_BASE_URL and headers of the requests. Refer to this file `./src/utils/axios.js` with content  

`const token = localStorage.getItem("token");
const DEV_BASE_USER = "http://localhost:3000/api/v1";
const PROD_BASE_USER = "https://titan-devs-staging.herokuapp.com/api/v1";
export default axios.create({
  baseURL:
    process.env.NODE_ENV === "production" ? PROD_BASE_USER : DEV_BASE_USER,

  headers: {
    Authorization: `Bearer ${token}`
  }
})`

File: src/views/Profile.js [hunk 5]
```diff
@@ -0,0 +1,341 @@
+import React, { Component } from "react";
+import { connect } from "react-redux";
+import { withRouter } from "react-router-dom";
+import PropTypes from "prop-types";
+import { fetchCurrentUser as getCurrentUser } from "../redux/actions";
+import { Loading } from "../components/common/Spinner/Loading";
+import Confirm from "../components/common/Buttons/confirm";
```
Comment (YvesIraguha) [inline src/views/Profile.js]: This button is deprecated, you can replace with `BasicButton` created after.

File: src/views/EditProfile.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+  handleFormInput,
+  saveUpdatedUser,
+  uploadImage
+} from "../redux/actions";
+import ErrorMessage from "../components/common/Message/error";
+import SuccessMessage from "../components/common/Message/success";
+
+export class EditProfile extends Component {
+  componentDidMount() {
+    const { username } = this.props.match.params;
```
Comment (espoirMur) [inline src/views/EditProfile.js]: EsLint is warning here!
Consider using destructuring in this line!

**Review by YvesIraguha — APPROVED** [review]

**Review by YvesIraguha — CHANGES_REQUESTED** [review]
@niomwungeri-fabrice you did a great job, you did a great job. Here are some of the points I want to highlight:
- When I edited my profile picture, I didn't see the updated picture.
- When I edit the profile, I get a message that I edited successfully, but for better user experience, I believe that I should be redirected to my updated profile. 
- I left some comments on lines of code that may need refactoring.

**Review by espoirMur — DISMISSED** [review]

**Review by YvesIraguha — APPROVED** [review]

File: src/__tests__/__actions__/loginActions.test.js (hunk 4 above)
Comment (espoirMur) [reply src/__tests__/__actions__/loginActions.test.js]: plus one for these ideas...

File: src/helpers/helpers.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+export const fieldRemover = loggedInUser => {
```
Comment (espoirMur) [inline src/helpers/helpers.js:1]: can you find a better name for this file?
helpers.js is a bit confusing ...

**Review by espoirMur — COMMENTED** [review]
Nice job bro!
But this PR is lengthy, in my view, it could be split into 3 different pull requests: one for edit profile, another one for view profile, and another one for update profile or settings.
It could facilitate the reviewer to review it because it's very hard to review a PR with 53 files.

File: src/helpers/helpers.js (hunk 7 above)
Comment (niomwungeri-fabrice) [reply src/helpers/helpers.js:1]: @espoirMur Can you suggest one. Thanks

**Comment by niomwungeri-fabrice** [comment]
> Nice Job Fab! Here are some keys points to consider :
> 
> * While reviewing this PR I was not able to edit my profile. I was getting a blank page with the edit button disabled...
> * Can you consider putting a default picture in case a user doesn't have any?
> * It looks like I can edit my profile without being logged in, is it possible to restrict someone who is not logged to edit his profile?

Hi, @espoirMur Thanks for the feedback. I have applied your feedback. Thanks

**Comment by niomwungeri-fabrice** [comment]
> Nice job bro!
> But this PR is lengthy, in my view, it could be split into 3 different pull requests: one for edit profile, another one for view profile, and another one for update profile or settings.
> It could facilitate the reviewer to review it because it's very hard to review a PR with 53 files.

Hi @espoirMur
I think we share the same view. But remember we had problems with backlogs and that was the reason behind this. Thanks

File: .eslintrc.json [hunk 8]
```diff
@@ -92,4 +92,4 @@
       }
     ]
   }
-}
+}
```
Comment (niomwungeri-fabrice) [inline .eslintrc.json]: New line

File: .sample.env [hunk 9]
```diff
@@ -0,0 +1,5 @@
+CLOUD_NAME="*********"
+UNSIGNED_UPLOAD_PRESET="***********"
+CLOUDINARY_URL="*************"
+API_URL="************"
+TOKEN="*******"
```
Comment (niomwungeri-fabrice) [inline .sample.env]: new line

File: src/views/Profile.js [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+    }
+    if (error === "No user with that name") {
+      history.push(`/not_found`);
+    }
+    if (loggedInUser) {
+      if (loggedInUser.username !== username) {
+        return (
+          <img
+            className="loading"
+            src={require("../assets/img/403.png")}
```
Comment (niomwungeri-fabrice) [inline src/views/Profile.js]: use es6

File: src/views/EditProfile.js [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        params: { username }
+      }
+    } = this.props;
+    const { onFetchProfile } = this.props;
+    onFetchProfile(username);
+  }
+
+  handleChange = e => {
+    const { onInputChange } = this.props;
+    onInputChange({ field: e.target.name, value: e.target.value });
```
Comment (niomwungeri-fabrice) [inline src/views/EditProfile.js]: Destructure targets

File: src/views/EditProfile.js [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+      loading,
+      loggedInUser,
+      onClearMessage
+    } = this.props;
+    const {
+      match: {
+        params: { username }
+      }
+    } = this.props;
+    setTimeout(() => {
```
Comment (niomwungeri-fabrice) [inline src/views/EditProfile.js]: Call clearMessage function When needed

File: package.json [hunk 13]
```diff
@@ -37,10 +37,18 @@
...
+    "cloudinary": "^1.14.0",
+    "cloudinary-core": "^2.6.3",
+    "cloudinary-react": "^1.1.0",
+    "dotenv": "^8.0.0",
+    "dotenv-webpack": "^1.7.0",
+    "isomorphic-fetch": "^2.2.1",
     "jest": "^24.7.1",
     "jsonwebtoken": "^8.5.1",
+    "moxios": "^0.4.0",
+    "nock": "^10.0.6",
```
Comment (abayo-luc) [inline package.json]: put moxios and nock under devdependencies.

File: package.json [hunk 14]
```diff
@@ -51,14 +59,17 @@
     "react-rte": "^0.16.1",
     "redux": "^4.0.1",
     "redux-devtools-extension": "^2.13.8",
-    "redux-thunk": "^2.3.0"
+    "redux-mock-store": "^1.5.3",
+    "redux-thunk": "^2.3.0",
+    "sinon": "^7.3.1"
```
Comment (abayo-luc) [inline package.json]: move sinion under devdep

File: src/components/common/Dialog/Dialog.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    this.setState({ open: false });
+    const { deleteProfile } = this.props;
+    if (del) deleteProfile();
+  };
+
+  render() {
+    const { open } = this.state;
+    return (
+      <div>
+        <a onClick={this.handleClickOpen} className="danger">
```
Comment (abayo-luc) [inline src/components/common/Dialog/Dialog.js]: use button or Link, it is not recommended to use anchor tag without href.

File: src/components/common/Spinner/Loading.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+
+export const Loading = () => (
+  <img
+    className="loading"
+    src={require("../../../assets/img/circle-loading.gif")}
```
Comment (abayo-luc) [inline src/components/common/Spinner/Loading.js]: User ES6: ```import``` instead of ```require```

File: src/views/EditProfile.js [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+    const {
+      match: {
+        params: { username }
+      }
+    } = this.props;
+
+    if (!loggedInUser) {
+      return <img className="loading" src={forbidden} alt="Forbidden" />;
+    }
+    if (loggedInUser) {
```
Comment (abayo-luc) [inline src/views/EditProfile.js:63]: can you combine these two blocks of if statement into one block?  ( just to make it dry)

File: src/views/Settings.js [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+      password,
+      currentPassword,
+      onChangePassword,
+      profile: { username }
+    } = this.props;
+    const data = { newPassword: password, currentPassword };
+    onChangePassword(data, username);
+  };
+
+  handleChange = async () => {
```
Comment (abayo-luc) [inline src/views/Settings.js]: if you are not using ```await``` inside this method, remove the ```async``` keyword.

**Review by abayo-luc — CHANGES_REQUESTED** [review]

File: src/__tests__/__views__/EditProfile.test.js [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    });
+    test("onInputChange()", () => {
+      const wrapper = mount(<EditProfile {...props} />);
+      wrapper
+        .find(".input_field")
+        .first()
+        .simulate("change", { target: { value: "hello" } });
+      expect(props.onInputChange).toHaveBeenCalled();
+    });
+    // test("loggedInUser()", () => {
```
Comment (abayo-luc) [inline src/__tests__/__views__/EditProfile.test.js]: remove these comments

**Review by abayo-luc — APPROVED** [review]

File: src/__tests__/__components__/Messages/ErrorMessage.test.js [hunk 20]
```diff
@@ -0,0 +1,10 @@
+import React from "react";
+import { shallow } from "enzyme";
+import ErrorMessage from "../../../components/common/Message/error";
+
+describe("Loading", () => {
```
Comment (oesukam) [inline src/__tests__/__components__/Messages/ErrorMessage.test.js]: @niomwungeri-fabrice Are these tests related to the loading component? Why did you name the describe as `Loading`?

File: src/__tests__/__actions__/authActions.test.js [hunk 21]
```diff
@@ -65,7 +65,7 @@ describe("Login action creators", () => {
         }
       ];
 
-      moxios.stubRequest(`${process.env.API_BASE_URL}/users/login`, {
+      moxios.stubRequest(`${process.env.API_URL}/users/login`, {
```
Comment (oesukam) [inline src/__tests__/__actions__/authActions.test.js]: @niomwungeri-fabrice Can't you destruct API_URL?

File: src/__tests__/__components__/Dialog/Dialog.test.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+};
+describe("Dialog()", () => {
+  test("should trigger click event", () => {
+    const component = shallow(<Dialog {...props} />);
+    component
+      .find("BasicButton")
+      .first()
+      .simulate("click");
+    expect(props.deleteProfile).not.toHaveBeenCalled();
+  });
```
Comment (oesukam) [inline src/__tests__/__components__/Dialog/Dialog.test.js:20]: @niomwungeri-fabrice Empty line required to separate test cases.

File: src/__tests__/__components__/Dialog/Dialog.test.js [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+  test("should call deleteProfile function", () => {
+    const component = shallow(<Dialog {...props} />);
+    component.find("#delete-profile").simulate("click");
+    expect(props.deleteProfile).toHaveBeenCalled();
+  });
+  test("should call handleClickOpen function", () => {
+    const component = shallow(<Dialog {...props} />);
+    component.find(".danger").simulate("click");
+    expect(props.open).toBe(true);
+  });
```
Comment (oesukam) [inline src/__tests__/__components__/Dialog/Dialog.test.js:32]: @niomwungeri-fabrice Empty line required.

File: src/__tests__/__components__/Messages/SuccessMessage.test.js [hunk 24]
```diff
@@ -0,0 +1,10 @@
+import React from "react";
+import { shallow } from "enzyme";
+import SuccessMessage from "../../../components/common/Message/success";
+
+describe("Loading", () => {
```
Comment (oesukam) [inline src/__tests__/__components__/Messages/SuccessMessage.test.js]: @niomwungeri-fabrice Why `Loading`?

File: src/__tests__/__helpers__/helpers.test.js [hunk 25]
```diff
@@ -0,0 +1,13 @@
+import { fieldRemover, nullRemover } from "../../helpers/helpers";
+import { initialState } from "../testData";
+
+describe("helpers", () => {
+  test("should remove nulls", () => {
+    const filteredObject = nullRemover(initialState);
+    expect(filteredObject).toHaveProperty("email");
+  });
```
Comment (oesukam) [inline src/__tests__/__helpers__/helpers.test.js:8]: Empty line required.

File: src/__tests__/__reducers__/userReducers.test.js [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+    expect(
+      userReducer(INITIAL_STATE, {
+        type: SET_LOADING,
+        payload: true
+      })
+    ).toEqual({
+      ...INITIAL_STATE,
+      loading: true
+    });
+  });
```
Comment (oesukam) [inline src/__tests__/__reducers__/userReducers.test.js:27]: Empty line required.

File: src/components/common/AppBars/navBar.js [hunk 27]
```diff
@@ -1,7 +1,8 @@
 import React, { Component } from "react";
-import { Link } from "react-router-dom";
-import Logo from "../../../assets/img/quill-drawing-a-line.svg";
+import { connect } from "react-redux";
+import { Link, withRouter } from "react-router-dom";
 import UserIcon from "../../../assets/img/user.jpg";
+import Logo from "../../../assets/img/quill-drawing-a-line.svg";
 
 class Navbar extends Component {
   constructor(props) {
```
Comment (oesukam) [inline src/components/common/AppBars/navBar.js]: @niomwungeri-fabrice We don't need the `constructor` with the react's version that we are using

File: src/components/common/Dialog/Dialog.js [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+          <DialogActions>
+            <BasicButton
+              title="Cancel"
+              onClick={() => this.handleClose(false)}
+            />
+            <BasicButton
+              id="delete-profile"
+              title="Delete my account"
+              onClick={() => this.handleClose(true)}
+              style={{ color: "red", fontWeight: "bold" }}
```
Comment (oesukam) [inline src/components/common/Dialog/Dialog.js]: We only use dynamic css in JS. It should be moved to scss

File: src/components/common/Message/message.js [hunk 29]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+const Message = ({ title, classType, icons }) => (
+  <div className={classType}>
+    <i className={icons} />
+    {title}
+  </div>
+);
```
Comment (oesukam) [inline src/components/common/Message/message.js:9]: @niomwungeri-fabrice Empty line required.

**Review by oesukam — CHANGES_REQUESTED** [review]
@niomwungeri-fabrice Make sure to separate your units (functions and classes) by an empty line

File: src/__tests__/__actions__/authActions.test.js (hunk 21 above)
Comment (niomwungeri-fabrice) [reply src/__tests__/__actions__/authActions.test.js]: Hi @oesukam!
Those aren't my changes but I can change them.
Thanks

File: src/components/common/AppBars/navBar.js (hunk 27 above)
Comment (niomwungeri-fabrice) [reply src/components/common/AppBars/navBar.js]: Hi @oesukam!
Those aren't my changes but I can change them.
Thanks

**Review by oesukam — APPROVED** [review]
